src/io/config.cpp

```diff
@@ -306,13 +306,21 @@ void Config::CheckParamConflict() {
     }
   }
   // Check max_depth and num_leaves
+  if (num_leaves > kMaxNumLeaves) {
+    Log::Fatal("num_leaves (%d) is larger than maximum allowed leaves (%d)", num_leaves, kMaxNumLeaves);
+  }
+
   if (max_depth > 0) {
-    int full_num_leaves = static_cast<int>(std::pow(2, max_depth));
+    double full_num_leaves = std::pow(2, max_depth);
     if (full_num_leaves > num_leaves
         && num_leaves == kDefaultNumLeaves) {
       Log::Warning("Accuracy may be bad since you didn't set num_leaves and 2^max_depth > num_leaves");
     }
-    num_leaves = std::min(num_leaves, 2 << max_depth);
+
+    if (full_num_leaves < kMaxNumLeaves && full_num_leaves < num_leaves) {
```

Comment:
as `num_leaves <= kMaxNumLeaves`, do we need to check `full_num_leaves < kMaxNumLeaves` here ?

Reply:
Yes to prevent the overflow during the double to int cast

Reply:
Alternative I can make this `num_leaves  <= INT_MAX` if that makes this more clear

Reply:
@chris-smith-zocdoc  I'm sorry, but I think that `full_num_leaves < kMaxNumLeaves` is excess too, because the second part `full_num_leaves < num_leaves` is more strict due to that we have a guarantee `num_leaves <= kMaxNumLeaves` (from `config.h`). Can you please elaborate a little bit more why it's needed?

Reply:
Ah yes, you are correct. I'll make the change